content/blog/architecture-as-code-microservices/index.md

```diff
@@ -0,0 +1,105 @@
+
+```ts
+export class Vpc extends ComponentResource {
+    vpc: aws.ec2.Vpc;
+    privateZone: aws.route53.Zone;
+    dhcpOptionSet: aws.ec2.VpcDhcpOptions;
+    internetGateway: aws.ec2.InternetGateway;
+    publicSubnets: aws.ec2.Subnet[] = [];
+    privateSubnets: aws.ec2.Subnet[] = [];
+    publicRouteTable: aws.ec2.RouteTable;
+    privateRouteTables: aws.ec2.RouteTable[] = [];
+    natGateways: aws.ec2.NatGateway[] = [];
+    natElasticIpAddresses: aws.ec2.Eip[] = [];
+
+   // ...
+
+`"
+
+The resource needs an interface to set the name and parameters for our new VPC class.
+
+```ts
+export interface VpcArgs {
+    description: string;
+    baseTags: aws.Tags;
+
+    baseCidr: string;
+    availabilityZoneNames: string[];
+    enableFlowLogs?: boolean;
+
+    endpoints: {
+        s3: boolean;
+        dynamodb: boolean;
+    };
+}
+```
+
+Finally, we need to be able to peer VPCs so that the services can communicate with each other. We can take advantage of TypeScript's [member function](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) to implement peering.
+
+```ts
+public configurePeering(args: PeerToArgs) {
+
+  // …
+}
+```
+
+We can reuse this pattern to build a reusable [database component](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/database/src/database.ts) from RDS. Likewise, the [REST application](https://github.com/pulumi/examples/blob/master/aws-stackreference-architecture/application/src/application.ts) creates a service using a container deployed in Fargate and fronted by an ALB. You can replace the container with an updated application without interrupting the service.
+
+## Jumping In
+
+We encourage you to try out deploying microservices with stacks. To download just the stack example from the  [examples repository](https://github.com/pulumi/examples/) we'll use a git sparse checkout to only copy the [aws-stackreference-architecture](https://github.com/pulumi/examples/tree/master/aws-stackreference-architecture) directory.
+
+```bash
+$ mkdir microservices
+$ cd microservices
+
+$ git init
+$ git remote add origin -f https://github.com/pulumi/examples/
+$ git config core.sparseCheckout true
+$ aws-stackreference-architecture >> .git/info/sparse-checkout
+$ git pull origin master
+```
+
+Follow the instructions to deploy the application. One caveat, when deploying the network microservice, set the number of availability zones to a minimum of two, e.g., `pulumi config set azCount 2`. Experiment with the example by adding data to the database and adding your REST application as a container. Add a web client to submit requests and display them. Let us know if you have questions on [Community Slack](https://pulumi-community.slack.com/) or send a tweet to show off your application.
```

Comment:
```suggestion
Follow the instructions to deploy the application. Experiment with the example by adding data to the database and adding your REST application as a container. Add a web client to submit requests and display them. Let us know if you have questions on [Community Slack](https://pulumi-community.slack.com/) or send a tweet to show off your application.
```


The `azCount` defaults to 2 in the networking stack:
https://github.com/pulumi/examples/blob/90f6fa541f7d6cf0b5171b39e7ef234b785cc98f/aws-stackreference-architecture/networking/src/index.ts#L7